src/test/java/bowlingexercise3/BowlingTest.java

```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
+    public void boundaryConditionsSpareBonusBallTest(String turns) {
+        Bowling game = new Bowling();
+        Assertions.assertDoesNotThrow(() -> game.turns(turns));
+    }
+
+
+    @ParameterizedTest
+    @CsvSource(value = {
+
+            "X|//////|X|45|8/|X|X||X||9/X",
+            "X|///|X|45|8/|X|X||X||9/X",
+            "X|XXXX|X|45|8/|X|X||X||9/X",
+            "X|X/|X|45|8/|X|X||X||9/X",
+            "X||X|45|8/|X|X||X|9/X",
+            "X|81|X|45|8/|X|X||X|9/X",
+            "11|11|11|11|11|/1|11|11|11",
+            "11|11|11|11|11|1X|11|11|11",
+            "1111888|11|11|11|11|11|11|11|11",
+            "1X?X|11|11|11|11|11|11|11|11",
+            "00|11|11|11|11|11|11|11|11|11||",
+            "10|11|11|11|11|11|11|11|11|11||",
+            "11|11|11|11|11|11|11|11|11|1//||",
+            "H1|11|11|11|11|11|11|11|11|1/||",
+            "11|11|11|11|11|11|11|11|1X2|11||",
+            "11|11|11|11|11|11|11|11|11|1X2||",
+            "11|11|11|11|11|11|11|11|11|1X||/",
+            "11|/1|11|11|11|11|11|11|11|1/||/",
+            "11|11|11|11|11|11|11|11|11|/1||",
+            "11|11|11|11|11|11|11|11|11|1X||",
+            "11|11|11|11|11|11|11|11|11|X/||1",
+
+    })
+
+    public void testInvalidInput(String turns) {
+
+        Bowling game = new Bowling();
+        Exception e = assertThrows(IllegalArgumentException.class, () -> game.turns(turns));
+        assertEquals("Input isn't valid", e.getMessage());
+    }
+
+    @ParameterizedTest
+    @EmptySource
+    @CsvSource(value = {
+
+
+            "/",
+            "|",
+            "||||||||||",
+            "------------------",
+            "1|11|11|11|11|11|11|11|11|11||",
+            "a1|11|11|11|11|11|11|11|11|11||",
+            "1a|11|11|11|11|11|11|11|11|11||",
+            "/2|11|11|11|11|11|11|11|11|11||",
+            "11|11|11|11|11|11|11|11|11||",
+            "11|11|11|11|11|11|11|11|11|11|11||",
+            "11|%1|11|11|11|11|11|11|11|11||",
+            "1+|11|11|11|11|11|11|11|11|11||",
+            "111|11|11|11|11|11|11|11|11|11||",
+            "11|11|11|11|11|11|11|11|11|111||",
+            "11|11||11|11|11|11|11|11|11|11|||",
+            "11|11|11|11|11|1X|11|11|11|11||",
+            "X1|11|11|11|11|11|11|11|11|11||",
+            "11|11|10|11|11|11|11|11|11|11||",
+            "//|11|11|11|11|//|11|11|11|11||",
+            "11|11|11|11|11|11|11|11|-X|11||",
+            "///|11|22|11|11|//|11|11|11|11||",
+            "11|11|45|11|11|11|11|11|-X|11||",
+            "X|X|X|X|X|X|X|X|X|XX||",
+            "X|X|X|X|X|X|X|X|X|XXX||",
+            "5|X|X|X|X|X|X|X|X|X||",
+            "X|X|X|X|X|X|X|X|X|/||",
+            "X|X|X|X|X|X|X|X|X|6||XX",
+            "X|X|X|X|X|X|X|X|X|?||35",
+            "X|X|X|X|X|X|X|X|X|X||81-",
+            "9|X|X|X|X|X|2X|X|X|X||",
+            "9|X|X|X|X|X|X|X|X|X||",
+            "/X|X|X|X|X|X|X|X|X|X||",
+            "-/|X|X|X|X|X|X|X|X|X||-X",
+
+
+    })
+    public void TestingInvalidInput(String turns) {
```

Comment:
I get confused between this method and the previous one? So there are some tests for the wrong number of frames, invalid characters in frames 1-9, frame 10 and bonus balls?

Reply:
Yes, I have separated tests into frame sizes, with/without bonus frames, and other variations

Reply:
Could the test methods have different names or be split into groups? It's hard to tell what their purpose is as there are two methods with the same names?

Reply:
Have labelled it more clearly now